### PR TITLE
Output error when 'goal node stop' is run on invalid directory

### PR DESCRIPTION
**Pull request by nicholasguoalgorand**
## Summary

Output error when 'goal node stop' is run on invalid directory, gives user feedback instead of saying node successfully stopped.

## Test Plan

Ran the commands in terminal

### CODE REVIEW

**Review by tsachiherman — CHANGES_REQUESTED** [review]
Few quests:
1. add expect test to prove the correctness of your change.
2. The existing pattern of `alreadyStopped bool, err error` is common. If err is set, the alreadyStopped is expected to be ignored. Unfortunately, this isn't what we're after. We need to return here few possible errors that would need to be handled differently on the client side. My suggestion:

a. create an error struct for alreadyStopped.
b. create an error struct for invalidDataDirectory
c. change StopAlgod so that it would return only a single error, and add error testing & formatting upstream.

File: nodecontrol/algodControl.go [hunk 1]
```diff
@@ -133,7 +149,11 @@ func (nc NodeController) algodRunning() (isRunning bool) {
 }
 
 // StopAlgod reads the net file and kills the algod process
-func (nc *NodeController) StopAlgod() (alreadyStopped bool, err error) {
+func (nc *NodeController) StopAlgod() (err error) {
+	// Check for valid data director
+	if !util.IsDir(nc.algodDataDir) {
```
Comment (algonautshant) [inline nodecontrol/algodControl.go:154]: This comment is for something not part of this PR, but I would like to raise it here:
Why do we say "kill node"? Shouldn't it be "stop node"? 
```
 $ goal node stop -d /tmp/a.txt 
Cannot kill node: /tmp/a.txt is not a valid directory
```
Instead:
`Cannot stop node: /tmp/a.txt is not a valid directory.`

**Review by algonautshant — COMMENTED** [review]
Looks good. 
I have one comment and one (unrelated) comment.

File: nodecontrol/algodControl.go [hunk 2]
```diff
@@ -40,6 +40,22 @@ const StdErrFilename = "algod-err.log"
 // StdOutFilename is the name of the file in <datadir> where stdout will be captured if not redirected to host
 const StdOutFilename = "algod-out.log"
 
+type NodeAlreadyStoppedError struct {
```
Comment (algonautshant) [inline nodecontrol/algodControl.go:43]: Travis failure: Comment is missing here.
`nodecontrol/algodControl.go:43:6: exported type NodeAlreadyStoppedError should have comment or be unexported
`

File: nodecontrol/algodControl.go [hunk 3]
```diff
@@ -40,6 +40,22 @@ const StdErrFilename = "algod-err.log"
...
+	algodDataDir string
+}
+
+func (e *NodeAlreadyStoppedError) Error() string {
+	return fmt.Sprintf("Node in directory %s has already stopped", e.algodDataDir)
+}
+
+type InvalidDataDirError struct {
```
Comment (algonautshant) [inline nodecontrol/algodControl.go:51]: Travis failure: Comment is missing here.

`nodecontrol/algodControl.go:51:6: exported type InvalidDataDirError should have comment or be unexported`

File: nodecontrol/algodControl.go [hunk 4]
```diff
@@ -40,6 +40,22 @@ const StdErrFilename = "algod-err.log"
 // StdOutFilename is the name of the file in <datadir> where stdout will be captured if not redirected to host
 const StdOutFilename = "algod-out.log"
 
+type NodeAlreadyStoppedError struct {
+	algodDataDir string
+}
+
+func (e *NodeAlreadyStoppedError) Error() string {
+	return fmt.Sprintf("Node in directory %s has already stopped", e.algodDataDir)
```
Comment (tsachiherman) [inline nodecontrol/algodControl.go:48]: nit: place the %s in a single-quotes, so that the user would be able to distinguish between the messages and the argument.

File: nodecontrol/algodControl.go [hunk 5]
```diff
@@ -40,6 +40,22 @@ const StdErrFilename = "algod-err.log"
...
+}
+
+type InvalidDataDirError struct {
+	algodDataDir string
+}
+
+func (e *InvalidDataDirError) Error() string {
+	return fmt.Sprintf("%s is not a valid directory", e.algodDataDir)
```
Comment (tsachiherman) [inline nodecontrol/algodControl.go:56]: This is not the "correct" error for this use case. using util.IsDir would return true if the directory exists.
It doesn't tell you if a directory of that name could be created.
In that case, you want a "MissingDataDirError" which would format into
"the provided data directory '%s' does not exists"
You could get more information by calling os.Stat and getting the precise underlying error.

File: nodecontrol/algodControl.go [hunk 6]
```diff
@@ -133,7 +149,11 @@ func (nc NodeController) algodRunning() (isRunning bool) {
 }
 
 // StopAlgod reads the net file and kills the algod process
-func (nc *NodeController) StopAlgod() (alreadyStopped bool, err error) {
+func (nc *NodeController) StopAlgod() (err error) {
+	// Check for valid data director
```
Comment (tsachiherman) [inline nodecontrol/algodControl.go:153]: nit: typo : director -> directory

File: nodecontrol/algodControl.go (hunk 1 above)
Comment (tsachiherman) [reply nodecontrol/algodControl.go:154]: yes, I agree that the user-facing error message should refrain of using the "kill" terminology, since it's the signal name and not the operation.

File: nodecontrol/algodControl.go [hunk 7]
```diff
@@ -143,8 +163,7 @@ func (nc *NodeController) StopAlgod() (alreadyStopped bool, err error) {
 			return
 		}
 	} else {
-		err = nil
-		alreadyStopped = true
+		return &NodeAlreadyStoppedError{algodDataDir: nc.algodDataDir}
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply nodecontrol/algodControl.go:166]: I think it's ok, but would drop the "configuration" and add single quotes around the directory name.

**Review by tsachiherman — CHANGES_REQUESTED** [review]
requested few changes. looks pretty good.

File: nodecontrol/algodContol_test.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package nodecontrol
```
Comment (tsachiherman) [inline nodecontrol/algodContol_test.go:17]: wait! I'm confused - is that an empty file ?!

File: nodecontrol/algodControl.go (hunk 1 above)
Comment (tsachiherman) [reply nodecontrol/algodControl.go:156]: @algonautshant - could you open another PR for the kill vs stop ?
it's in goal,
```
errorKill                         = "Cannot kill node: %s"
```
and would affect some of the changes made here ( but maybe more )

File: nodecontrol/algodContol_test.go (hunk 8 above)
Comment (nicholasguoalgorand) [reply nodecontrol/algodContol_test.go]: strange, git and renaming files work strange

File: test/e2e-go/cli/algod/expect/algodTelemetryLocationTest.exp [hunk 9]
```diff
@@ -19,6 +19,7 @@ if { [catch {
     #allows script to be run outside of go context
     exec mkdir -p $TEST_PRIMARY_NODE_DIR
 
+    exec goal node start -d $TEST_PRIMARY_NODE_DIR
```
Comment (tsachiherman) [inline test/e2e-go/cli/algod/expect/algodTelemetryLocationTest.exp:22]: I'm looking on this one and I'm wondering why we need to make the call to `goal node stop` ?
Can you try to run it without it to see if the test passes ?

**Review by tsachiherman — APPROVED** [review]
